src/servenix/server/servenix.py

```diff
@@ -87,24 +112,37 @@ def store_path_from_hash(self, store_object_hash):
             self._hashes_to_paths.pop(store_object_hash, None)
             raise NoSuchObject("No object with hash {} was found!"
                                .format(store_object_hash))
-        # Get the list of store objects by listing the directory.
-        # Iterate through them until a matching hash is found, or
-        # we've exhausted all paths, in which case we error.
-        for path in map(decode_str, os.listdir(self._nix_store_path)):
-            match = _PATH_REGEX.match(path)
-            if match is None:
-                continue
-            path = join(self._nix_store_path, path)
-            prefix = match.group(1)
-            # Add every path seen to the _hashes_to_paths cache.
-            self._hashes_to_paths[prefix] = path
-            if prefix == store_object_hash:
-                # The path exists in the store. Ensure it's also a
-                # valid path according to nix-store.
-                if not self.check_in_store(path):
-                    break
+        if self._db_con is not None:
+            # If we have a direct database connection, use this to check
+            # path existence.
+            query = ("select path from ValidPaths where path like '{}%'"
+                     .format(join(self._nix_store_path, store_object_hash)))
+            with self._db_con:
+                paths = self._db_con.execute(query).fetchall()
```

Comment:
Alright, working example is something like this.  The wildcard character needs to be in the value that you're subbing in:
```
query = "select path from ValidPaths where path like %s"
wildcard_path = join(self._nix_store_path, store_object_hash) + "%"
with self._db_con:
    paths = self._db_con.execute(query, wildcard_path).fetchall()
```